### ISSUE TITLE
Uploading files via API - URL: Missing value error

### ISSUE THREAD

**Issue by KrzysztofMadejski** [ckan/ckan#2263]
I'm trying to to upload file via CKAN API using Python api: https://github.com/ckan/ckanapi#file-uploads

The problem is I'm getting {u'url': [u'Missing value'], u'__type': u'Validation Error'}
CKAN 2.3

Here's resource I'm uploading:
resource_create: {"name": "Statystyki strukturalne maj 2014", "created": "2014-07-01T13:45:08", "package_id": "537b245ebdd0a48259924918", "upload": "/tmp/infona/5486df35bdd0789a545de155/Statystyki strukturalne maj 2014.zip", "state": "active", "last_modified": "2014-12-09T11:38:29", "revision_id": 2, "id": "53b2bb64bdd0a4825992535b"}

And code responsible for uploading:
r.update({
                    'upload': open(local_path)
                })
ckan.action.resource_create(**r)

I guess that before validation file should be stored locally and that value should go into URL. But it should all be done internally. This scenario works perfectly from web interface.

**Edited**
Adding empty string as 'url' solves the issue, but it's really counterintuitive. I believe API should accept either 'upload' field or 'url' but not both at the same time.

**Comment by wardi** [comment]
:+1: let's change the [resource url field validation](https://github.com/ckan/ckan/blob/master/ckan/logic/schema.py#L81) from:

```python
'url': [not_empty, unicode, remove_whitespace],
```

to something like 

```python
'url': [ignore_if_file_passed, not_empty, unicode, remove_whitespace]
```

**Comment by KrzysztofMadejski** [comment]
I like this idea!

So pseudocode for `ignore_if_file_passed` would be:
- if file passed
  - if url is non-empty throw validation error: Cannot specify both 'url' and 'upload'
  - else throw StopOnError (ignore url)
- else pass

Are we sure that file will be processed before this validator is fired?

BTW: (I couldn't find it in documention). How one should create validator that is checking few fields at once (relationships between fields)?

**Comment by wardi** [comment]
http://docs.ckan.org/en/latest/extensions/adding-custom-fields.html#custom-validators validator functions that take four parameters give you all the data so you can check other fields passed

**Comment by KrzysztofMadejski** [comment]
I've tried solution with ignore_if_file_passed and it didn't work. Here's the thing:
When I upload a file, resource['url'] will get populated with filename by `upload = uploader.ResourceUpload(data_dict)`.
That is perfectly ok, but before we reach that point we get busted by `_get_or_bust(data_dict, 'url')`. 

So first thing is to drop `_get_or_bust(data_dict, 'url')` line. On upload url will be autopopulated anyways.

Second thing is responding to wrong input. If there's no upload nor url one would expect `url.not_empty` validator to be fired. But right now `if_empty_guess_format` assumes url is always set and is 500ing on `None.some_func`. So I've added `if url:` assuming it's not this validator thing to check url presence.

Please cherrypick. :)

**Comment by wardi** [comment]
@KrzysztofMadejski Would you submit a PR with da0b1ae and a test to reproduce the issue?

**Comment by stevage** [comment]
Any update on this? Just ran into this same problem using httpie (as suggested in the [docs](http://docs.ckan.org/en/latest/api/#example-importing-datasets-with-the-ckan-api)):
```
 http --json POST http://ckanau.org/api/3/action/resource_update id=9036d9f5-eee4-402e-b837-c12136666524 upload=@treesheader.csv Authorization:my-key
```
```http
HTTP/1.1 409 Conflict
Cache-Control: no-cache
Connection: close
Content-Length: 158
Content-Type: application/json;charset=utf-8
Date: Thu, 26 Nov 2015 01:24:21 GMT
Pragma: no-cache
Server: Apache/2.2.31 (Amazon)
```
```json
{
    "error": {
        "__type": "Validation Error",
        "url": [
            "Missing value"
        ]
    },
    "help": "http://ckanau.org/api/3/action/help_show?name=resource_update",
    "success": false
}
```

Or if there was any workaround I'd be extremely happy. I'm trying to give advice on automatic publishing to several government ckan's, so it could be a long time before any bug fix was implemented in all of them.

**Comment by stevage** [comment]
Actually, this seems to be ok for me, as long as I pass a `format` attribute as well:

`ckan.action.resource_update(id=args.resource,upload=open(args.filename), format="csv")`

It still doesn't work using httpie though:

```
http --json POST http://ckanau.org/api/3/action/resource_update id=9036d9f5-eee4-402e-b837-c12136666524 format=csv upload=@newtrees.csv Authorization:mykey
```

Adding in `url=" "` returns a 200 HTTP code and `{ ... "success": true }`, but the resource doesn't actually get updated.

**Comment by KrzysztofMadejski** [comment]
I didn't have time to write test for it so it's still hanging out there.

**Comment by stevage** [comment]
In case it's relevant, here's my code that uses this API: 

https://github.com/OpenCouncilData/Ckan-Upload/blob/master/ckan-upload.py

**Comment by Kyle-Falconer** [comment]
I'm also experiencing this issue, but passing in `url=''` fixes it for me for now.

**Comment by wardi** [comment]
fixed by #2844
